rfc-117-containerised-runtime.md

```diff
@@ -0,0 +1,68 @@
+# RFC 000: Use a containerised runtime
+
+## Summary
+
+Use a containerised runtime for hosting GOV.UK applications. GOV.UK applications currently run on virtual machines running Ubuntu Trusty Tahr. There is currently an extended support contract for Ubuntu Trusty which will end in April 2022. Rather than upgrading Ubuntu on the virtual machines GOV.UK should change the application runtime to a containerised platform.
+
+## Problem
+
+GOV.UK currently hosts the majority of its applications and services using virtual machines running Ubuntu 14.04 LTS (Trusty Tahr).
+
+Trusty is out of date, it was released in April 2014 and standard support ended in April 2019. At this time GOV.UK signed up for [Extended Security Maintenance](https://ubuntu.com/esm) (ESM) which provides continued security fixes for high and critical common vulnerabilities and exposures (CVEs) for the [most commonly used packages](https://wiki.ubuntu.com/SecurityTeam/ESM/14.04#A14.04_Infrastructure_ESM_Packages). ESM will end in April 2022.
+
+Running an old operating system is preventing GOV.UK from upgrading key components of our technology stack (MongoDB, Puppet, Python, ...).
+
+Upgrading Ubuntu in the existing infrastructure is difficult for a number of reasons, including but not limited to:
+
+- Multiple applications share the operating system of the same virtual machine. This means it’s not easy to upgrade a single application at a time. Furthermore, the way our [Puppet code](http://github.com/alphagov/govuk-puppet/) is written makes supporting virtual machines on different operating systems at the same time hard to support.
+- GOV.UK currently uses upstart as our init system, but newer versions of Ubuntu switch to systemd. This means GOV.UK will need to do a significant rewrite of our init scripts as part of an upgrade.
```

Comment:
Presumably moving to a containerised system requires an even greater rewrite/rethink of these than the switch to systemd? so this may be a problem in all scenarios.

Reply:
Do you have an example @kevindew? I assume whatever container management system we use would essentially replaces upstart (which is probably out of scope for now)

Reply:
I think Kevin's point is that switching from Upstart to a "container runtime" for process management is a bigger switch than switching from Upstart to systemd, as systemd is more similar to upstart than container runtimes, as that's a different but related concept.

Reply:
Yup, @cbaines summarised it well 👍

Reply:
Yeah, the container runtime does replace the init system, but it's true that we'll still need to do the work to make sure that the applications behave reliably in the new world. I guess this work will be quite similar in nature to the switch to systemd, but at least we won't have to do it with puppet 😅.

I still think it's worth calling this out as a thing that makes upgrading past Trusty difficult, even though the solution in the RFC doesn't make the work go away entirely.

Reply:
> GOV.UK currently uses upstart as our init system, but newer versions of Ubuntu switch to systemd. This means GOV.UK will need to do a significant rewrite of our init scripts as part of an upgrade.

I agree with Kevin and Chris's points. It might be useful to have an example in the RFC to demonstrate how significant rewriting these scripts would be.

Reply:
I'm not sure I can actually support this statement. In theory puppet supports both upstart and systemd, and it should magically handle it all for us. There are a [few references to upstart specifically in govuk-puppet](https://github.com/alphagov/govuk-puppet/search?q=upstart&unscoped_q=upstart), but not so many that it looks like it would be a massive job.

In practice I think it would be a bit harder than that, but without trying it I can't know for sure. I absolutely do not want to try it myself 🙅‍♂.

Maybe let's just remove this point since it's controversial?

Reply:
I've watered the point down, rather than remove it, in af3889bda2efd9c68a8a6c6f1ab05c67e3c5c293.

I don't know enough about the systemd upgrade to come up with concrete examples of why it would be hard, but my spidey senses are definitely tingling with danger.